### ISSUE TITLE
Website issue: Laws are missing for farming-related use cases

### ISSUE THREAD

**Issue by TatharNuar** [StrangeLoopGames/EcoIssues#5059]
Server, Client or Website? Website
OS/Web Browser? Windows 10, Mozilla Firefox 58.0.2
Client Version? Beta 7.1.2
Detailed Description of Issue: See below
Reproducible? How difficult is it to reproduce? Yes, affects everyone
Steps to reproduce bug/issue: Attempt to propose a law
Additional Comments: See below

In 7.1.2, there is no option to create laws for when someone plows a field or when someone plants trees or crops on a field.  Additionally, the Harvest action doesn't have a way to differentiate between wild crops and domesticated crops planted by players.

As an example, if someone wants to allocate government funds whenever someone plants a tree in a particular district, but tax when people chop down wild trees in the same district without taxing when planted trees are chopped down, neither use case can be proposed as a law.

**Comment by LiaungYip** [comment]
Suggest that the below player actions/consequences be trackable.

Some should be trackable in Laws, others in Contracts, and some in both Laws and Contracts.

----

* Crops
  * Till soil (with the Hoe tool)
  * Fertilize soil
  * Plant seed (#4720 / #4077)
  * Clean up dead plant
  * Remove weeds from tilled soil (i.e. grass)
  * Harvest _immature_ plant, in the _wild_
  * Harvest _mature_ plant, in the _wild_
  * Harvest _immature_ plant, on _farmland_
  * Harvest _mature_ plant, on _farmland_

Example 1: Issue a contract to - 
1. Harvest `Mature Tomato Plants` from property `Tomato Farm` and deliver produce to `Tomato Chest`, 
2. Remove all weeds from property `Tomato Farm`,
3. Plant `50 Tomato Seed` on property `Tomato Farm`,
4. Fertilize soil on `Tomato Farm` to achieve Nitrogen level at least 70%.
5. Require the Contractor to have a minimum `Forest Forager` skill of 3. (#4291)

Example 2: Propose a law that discourages harvesting of wild plants, by taxing $1 for each wild plant harvested.

----

* Making trees
  * Plant tree seed (#4720 / #4077)
  * Tree planted *by player* grows into sapling
  * Tree planted *by player* grows into mature tree

* Harvesting trees
  * Chop seedling
  * Chop down immature tree
  * Trim leaves off tree
  * Trim branch off tree
  * Chop down mature tree
  * Create tree debris. (Charge per tree debris created.)
  * Remove tree debris (Already available in Laws and Contracts.)
  * Remove tree stump (Available in Contracts - not available in Laws. #4722.)

Example 1: Encourage planting trees, by allocating money when a player-planted tree grows into a semi-mature tree, and again when a player-planted tree grows into a fully-grown tree. _Note: Simply allocating money to plant a tree seed isn't sufficient - it's very easy to plant a dense grid of tree seeds, that will never grow into mature trees._

Example 2: Encourage players to gather tree seeds, by allocating money when they chop at tree leaves.

Example 3: Encourage players to clean up after chopping trees down, by making a law that taxes the creation of tree debris and tree stumps, and refunds the money if those are cleaned up.

**Comment by TatharNuar** [comment]
That comment got me thinking about how a solution to this issue would work.  For the sake of simplicity between the two systems, would it make more sense to just treat laws and contracts the same way, and just make a contract like a law that only applies to the people who opt into it?

**Comment by MarjieAVolk** [comment]
Partial duplicate of https://github.com/StrangeLoopGames/EcoIssues/issues/4077, as stated above.  Thanks for the examples, will keep this in mind the next time I work on the law system.

**Comment by LiaungYip** [comment]
> would it make more sense to just treat laws and contracts the same way, and just make a contract like a law that only applies to the people who opt into it?

The main difference between Laws and Contracts (as it stands) are that Laws can tax/allocate an amount of money *per action*, where as Contracts are *lump sum* - you get the same amount of money for fulfilling a contract, whether you performed zero actions (!), one action or 1,000 actions.

Personally, I wouldn't mind the  ability to have Contracts that paid out an amount per-item or per-action. This would allow players to post repeating contracts for tasks like "Mow Lawn" or "Harvest Field". 

Currently, it's not wise to post repeating contracts for those tasks. I haven't tested this, but I believe an evil Contractor can bankrupt a Principal who offers a repeating "Harvest Field" contract, like so:

1. Principal offers a repeating "Harvest Field" contract, paying $50 for each completion of the contract.
2. Evil Contractor takes the contract, harvests the field, and is paid $50.
3. The field now contains zero harvestable plants.
4. Evil Contractor immediately takes a repeat of the same contract.
5. There are no harvestable plants in the field, so the Evil Contractor can immediately complete the contract, for another $50.
6. Evil Contractor repeats the cycle until the Principal has $0.

A slightly less evil version of this goes like:

1. Principal offers a repeating "Harvest Field" contract for $100.
2. Ignorant Contractor takes the contract, harvests 100 mature Wheat from the field, and puts the Wheat into the nominated Storage Chest. (All is well and good.)
3. Principal re-plants the field with Wheat Seeds, expecting that they will be allowed to grow to 100%.
4. A few hours pass, and the Wheat grows to 50% maturity.
5. Ignorant Contractor takes a repeat of the contract, harvesting all the immature Wheat plants, producing only 50% of the Wheat.
6. Ignorant Contractor completes the contract and is paid $100, despite producing only half the output.

With some improvements, the Contracts system could support repeating contracts like the above.

* If a Contract offered payment per item or per action, the Evil Contractor would only be paid for actual work performed.
* If a Contract only allowed harvesting of mature plants, the Ignorant Contractor would fall into the [pit of success](https://blog.codinghorror.com/falling-into-the-pit-of-success/) - they would only be able to harvest plants which will give 100% yield.

**Comment by MarjieAVolk** [comment]
Okay all, based on your suggestions and requests, I have added actions "Plant", "Remove Stump", and "Harvest Leaves".  I decided not to go with "Plow" at this time, but I ensured that plowing or digging up a plant will count as a "Harvest".  I also did not add any conditions on the state of the plant being harvested, like maturity or wild/farmed, just for technical reasons, since that would require some bigger changes in order to support.  This will all be in 8.0.
